framework/src/modules/interoperability/mainchain/store.ts

```diff
@@ -33,6 +44,78 @@ export class MainchainInteroperabilityStore extends BaseInteroperabilityStore {
 		return true;
 	}
 
+	public async forward(ccmForwardContext: CCMForwardContext): Promise<void> {
+		const {
+			ccm,
+			eventQueue,
+			logger,
+			networkIdentifier,
+			getAPIContext,
+			getStore,
+		} = ccmForwardContext;
+		const apiContext = getAPIContext();
+		const tokenAPI = this.interoperableModuleAPIs.get(MODULE_ID_TOKEN) as TokenAPI | undefined;
+		const beforeCCMSendContext = createCCMsgBeforeSendContext({
+			ccm,
+			eventQueue,
+			getAPIContext,
+			logger,
+			networkIdentifier,
+			getStore,
+			feeAddress: EMPTY_FEE_ADDRESS,
+		});
+
+		if (!tokenAPI) {
+			return;
+		}
+
+		const receivingChainIDAsStoreKey = getIDAsKeyForStore(ccm.receivingChainID);
+		let receivingChainAccount;
+		try {
+			receivingChainAccount = await this.getChainAccount(receivingChainIDAsStoreKey);
+		} catch {
+			return;
+		}
```

Comment:
prefer this condition to return after we call bounce as suggested in the LIP,
its better we simply check everything early stage and then handle failed conditions later as in the LIP,
```ts
const chainAccount = await chainAccountExist(ccm.receivingChainID);
if (chainAccount && chainAccount.status === CHAIN_ACTIVE && isLive(ccm.receivingChainID)) {
try {
await token.forwardMessageFee(ccm)
} catch {// return }
addToOutbox()
return
}
```

Reply:
I think your snippet doesn't emphasize the real problem. In the LIP existence of chain account and the instance of the record is both denoted by `chainAccount(receivingChainID)`. But in our codebase we have two different methods for retrieving the method and checking for its existence. If we check the existence with `chainAccountExists()`, then we need to implement type predicate for that method but type predicates for promises are not implemented yet in Typescript. Then we need to set receivingChainAccount to `null` on error since if we don't catch when chain account retrieval method throws `NotFoundError` we are going to have control flow interruption before bounce. I will share my solution soon.